src/spec.c

```diff
@@ -1112,206 +1187,339 @@ static void IndexStats_RdbSave(RedisModuleIO *rdb, IndexStats *stats) {
+  IndexSpec *sp = scanner->spec_opt; // can be NULL
+
   RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
   RedisModuleScanCursor *cursor = RedisModule_ScanCursorCreate();
-  //  RedisModule_ThreadSafeContextLock(ctx);
-  while (RedisModule_Scan(ctx, cursor, IndexSpec_ScanCallback, NULL)) {
-    //    RedisModule_ThreadSafeContextUnlock(ctx);
-    //    RedisModule_ThreadSafeContextLock(ctx);
+  RedisModule_ThreadSafeContextLock(ctx);
+
+  if (sp) {
+    RedisModule_Log(ctx, "notice", "Scanning index %s in background", sp->name);
+  } else {
+    RedisModule_Log(ctx, "notice", "Scanning indexes in background");
   }
 
-  //  RedisModule_ThreadSafeContextUnlock(ctx);
-  RedisModule_ScanCursorDestroy(cursor);
+  if (!sp) {
+    Indexes_SetTempSpecsTimers();
+  }
 
-  // set the timers for temporary indexes:
-  dictIterator *iter = dictGetIterator(specDict);
-  dictEntry *entry = NULL;
-  while ((entry = dictNext(iter))) {
-    IndexSpec *sp = dictGetVal(entry);
-    if (!(sp->flags & Index_Temporary)) {
+  while (RedisModule_Scan(ctx, cursor, Indexes_ScanCallback, scanner)) {
+    if ((scanner->scannedKeys % REINDEX_BATCH_SIZE) >= 0) {
       continue;
     }
 
-    sp->timerId = RedisModule_CreateTimer(RSDummyContext, sp->timeout, IndexSpec_TimedOut, sp);
-    sp->isTimerSet = true;
+    if (sp) {
+      sp->keysIndexed += REINDEX_BATCH_SIZE;
+    } else {
+      dictIterator *iter = dictGetIterator(specDict);
+      dictEntry *entry = NULL;
+      while ((entry = dictNext(iter))) {
+        IndexSpec *sp = dictGetVal(entry);
+        sp->keysIndexed += REINDEX_BATCH_SIZE;
+      }
+      dictReleaseIterator(iter);
+    }
+
+    RedisModule_ThreadSafeContextUnlock(ctx);
+    sched_yield();
+    RedisModule_ThreadSafeContextLock(ctx);
+
+    if (sp && sp->isDropped) {
+      RedisModule_Log(ctx, "notice", "Scanning index %s in background: aborted - index dropped", sp->name);
+      sp->isReindexing = false;
+      sp->keysTotal = sp->keysIndexed = 0;
+      goto end;
+    }
   }
-  dictReleaseIterator(iter);
 
+  if (sp) {
+    sp->keysTotal = sp->keysIndexed = scanner->totalKeys;
+    sp->isReindexing = false;
+  } else {
+    dictIterator *iter = dictGetIterator(specDict);
+    dictEntry *entry = NULL;
+    while ((entry = dictNext(iter))) {
+      IndexSpec *sp = dictGetVal(entry);
+      sp->keysTotal = sp->keysIndexed = scanner->totalKeys;
+      sp->isReindexing = false;
+    }
+    dictReleaseIterator(iter);
+  }
+
+  RedisModule_Log(ctx, "notice", "Scanning indexes in background: done (scanned=%ld)", scanner->totalKeys);
+
+end:
+  RedisModule_ThreadSafeContextUnlock(ctx);
+  RedisModule_ScanCursorDestroy(cursor);
+
+  rm_free(scanner);
   RedisModule_FreeThreadSafeContext(ctx);
 }
 
-void IndexSpec_ScanAndReindex() {
-  /*if (!reindexPool) {
+//---------------------------------------------------------------------------------------------
+
+static void IndexSpec_ScanAndReindexAsync(IndexSpec *sp) {
+  if (sp->isDropped) {
```

Comment:
Again if it's dropped you can not access it, this is why you need a scan task struct.

Reply:
isDropped is "intending to drop". It isn't actually dropped until the indexing has finished/aborted.
PS we have a scan task struct.

Reply:
But then when you drop you need to wait and you might get into deadlock no?